modules/core/src/main/scala/fs2/kafka/internal/WithTransactionalProducer.scala

```diff
@@ -0,0 +1,63 @@
+/*
+ * Copyright 2018-2022 OVO Energy Limited
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+package fs2.kafka.internal
+
+import cats.effect.concurrent.Semaphore
+import cats.effect.{Concurrent, ContextShift, Resource}
+import cats.implicits._
+import fs2.kafka.internal.syntax._
+import fs2.kafka.{KafkaByteProducer, TransactionalProducerSettings}
+
+private[kafka] sealed abstract class WithTransactionalProducer[F[_]] {
+  def apply[A](f: (KafkaByteProducer, Blocking[F], ExclusiveAccess[F, A]) => F[A]): F[A]
+
+  def exclusiveAccess[A](f: (KafkaByteProducer, Blocking[F]) => F[A]): F[A] = apply {
+    case (producer, blocking, exclusive) => exclusive(f(producer, blocking))
+  }
+
+  def blocking[A](f: KafkaByteProducer => A): F[A] = apply {
+    case (producer, blocking, _) => blocking(f(producer))
+  }
+}
+
+private[kafka] object WithTransactionalProducer {
+  def apply[F[_], K, V](
+    settings: TransactionalProducerSettings[F, K, V]
+  )(
+    implicit F: Concurrent[F],
+    context: ContextShift[F]
+  ): Resource[F, WithTransactionalProducer[F]] =
+    (blockingResource(settings.producerSettings), Resource.eval(Semaphore(1))).tupled.flatMap {
+      case (blocking, accessSemaphore) =>
+        Resource[F, WithTransactionalProducer[F]] {
+          settings.producerSettings.createProducer.flatMap { producer =>
+            val withProducer = create(producer, blocking, accessSemaphore)
+
+            val initTransactions = withProducer.blocking { _.initTransactions() }
+
+            val close = withProducer.exclusiveAccess {
+              case (producer, blocking) =>
+                blocking(producer.close(settings.producerSettings.closeTimeout.asJava))
+            }
```

Comment:
I realised that closing the producer should probably have exclusive access too!

Reply:
Not sure whether it's necessary but definitely shouldn't do any harm

Reply:
Yes, I think you're right it might not be necessary, because the producer will wait for any in flight records to be flushed or the provided timeout to expire. From the perspective of not letting any other records get added to the producer during shutdown this is quite nice, but this shouldn't happen thanks to `Resource`, unless someone has done something silly like leak the producer out of the resource.

It also changes the behaviour of the `closeTimeout` setting for the transactional producer; we're using the semaphore to sequence transactions where we wait for the client buffer to flush and transaction comitted to the broker, the `closeTimeout` effectively becomes redundant when we use the semaphore with `close` because the buffer will always be empty. Given this do you think it's better to remove the `exclusiveAccess` call here?